lang/csharp/src/apache/main/AvroDecimal.cs

```diff
@@ -0,0 +1,983 @@
+        public AvroDecimal(byte[] value)
+        {
+            byte[] number = new byte[value.Length - 4];
+            byte[] flags = new byte[4];
+
+            Array.Copy(value, 0, number, 0, number.Length);
+            Array.Copy(value, value.Length - 4, flags, 0, 4);
+
+            UnscaledValue = new BigInteger(number);
+            Scale = BitConverter.ToInt32(flags, 0);
+        }
+
+        /// <summary>
+        /// Gets the unscaled integer value represented by the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public BigInteger UnscaledValue { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether the current <see cref="AvroDecimal"/> represents an even number.
+        /// </summary>
+        public bool IsEven
+        {
+            get { return UnscaledValue.IsEven; }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the current <see cref="AvroDecimal"/> represents the number '1'.
+        /// </summary>
+        public bool IsOne
+        {
+            get { return UnscaledValue.IsOne; }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the current <see cref="AvroDecimal"/> represents a number that is a power of two.
+        /// </summary>
+        public bool IsPowerOfTwo
+        {
+            get { return UnscaledValue.IsPowerOfTwo; }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the current <see cref="AvroDecimal"/> represents the number '0'.
+        /// </summary>
+        public bool IsZero
+        {
+            get { return UnscaledValue.IsZero; }
+        }
+
+        /// <summary>
+        /// Gets the sign of the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public int Sign
+        {
+            get { return UnscaledValue.Sign; }
+        }
+
+        /// <summary>
+        /// Gets the scale of the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public int Scale { get; }
+
+        /// <summary>
+        /// Converts the current <see cref="AvroDecimal"/> to a string.
+        /// </summary>
+        /// <returns>A string representation of the numeric value.</returns>
+        public override string ToString()
+        {
+            var number = UnscaledValue.ToString(CultureInfo.CurrentCulture);
+
+            if (Scale > 0)
+                return number.Insert(number.Length - Scale, CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator);
+
+            return number;
+        }
+
+        /// <summary>
+        /// Converts the current <see cref="AvroDecimal"/> to a byte array.
+        /// </summary>
+        /// <returns>The current <see cref="AvroDecimal"/> converted to an array of bytes.</returns>
+        public byte[] ToByteArray()
+        {
+            var unscaledValue = UnscaledValue.ToByteArray();
+            var scale = BitConverter.GetBytes(Scale);
```

Comment:
The spec says:

> The byte array must contain the two's-complement representation of the unscaled integer value in big-endian byte order. The scale is fixed, and is specified using an attribute.

This method is used by `ConvertToBaseValue` in `Decimal` to achieve this. However this method serializes the scale in addition to the unscaled value, so it doesn't conform.

Reply:
it'd be ideal to see some evidence of cross language integration testing. not sure how best to set that up, but maybe have a java test app that writes data to disk, read by a dotnet test app & vice versa.

Reply:
Good point, Matt. The [interop tests] exist to test some of the cross language features. However, the [`interop.avsc`] does not currently define any logical types.

@Fokko, are you aware of any interop tests that exercise logical types? Maybe I'm just missing something.

[interop tests]: https://github.com/apache/avro/tree/master/lang/csharp/src/apache/test/Interop
[`interop.avsc`]: https://github.com/apache/avro/blob/master/share/test/schemas/interop.avsc

Reply:
No, you're right. They are missing. However, if you're able to parse the primitive type of an `int` or `long`, you should also be able to parse the logical type.

Reply:
`Decimal.ConvertToBaseValue(...)` doesn't use `AvroDecimal.ToByteArray()` to serialize into an Avro payload. It gets the buffer from `AvroDecimal.UnscaledValue` (which is the underlying `BigInteger`). This is Avro specific encoding.

`AvroDecimal.ToByteArray()` on the other hand, returns a buffer that should be comparable with `Decimal.GetBits(...)`.